lib/file.c

```diff
@@ -651,27 +577,43 @@ static CURLcode file_do(struct Curl_easy *data, bool *done)
 
     if(!S_ISDIR(statbuf.st_mode)) {
       nread = read(fd, xfer_buf, bytestoread);
-    }
-    else {
-      memcpy(xfer_buf,
-             directory_data + already_read + data->state.resume_from,
-             bytestoread);
-      nread = bytestoread;
-      already_read += nread;
-    }
 
-    if(nread > 0)
-      xfer_buf[nread] = 0;
+      if(nread > 0)
+        xfer_buf[nread] = 0;
 
-    if(nread <= 0 || (size_known && (expected_size == 0)))
-      break;
+      if(nread <= 0 || (size_known && (expected_size == 0)))
+        break;
 
-    if(size_known)
-      expected_size -= nread;
+      if(size_known)
+        expected_size -= nread;
 
-    result = Curl_client_write(data, CLIENTWRITE_BODY, xfer_buf, nread);
-    if(result)
-      goto out;
+      result = Curl_client_write(data, CLIENTWRITE_BODY, xfer_buf, nread);
+      if(result)
+        goto out;
+    }
+    else {
+      DIR *dir = fdopendir(file->fd);
```

Comment:
You need to add a check for `fdopendir()` to configure and CMake so that we know it exists for the platform you build this. On Windows it won't exist. For now at least, I think we can presume the other POSIX dir funtions exist if this one does.

Then you can either opt to fail on other platforms, or continue to add code support other dir-reading functions. The Windows being a primary target I guess.

macOS also needs attention as the compiler there says:

**'fdopendir' has been marked as being introduced in macOS 10.10 here, but the deployment target is macOS 10.9.0**

Reply:
Hi @bagder,
Thanks for your feedback. I will add the configure/CMake checks this evening, and probably switch to opendir(file->path) to address the Mac OS X case. 
For Windows, I'll see if I can find someone with Windows experience to help me either add wrappers or #ifdefs.